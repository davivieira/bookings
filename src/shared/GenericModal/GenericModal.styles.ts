import { Modal, styled } from "@mui/material";

export const SCModal = styled(Modal)(({ theme: { palette } }) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  '& .modal-content': {
    backgroundColor: palette.background.default,
    outline: 'none',
  },
}))