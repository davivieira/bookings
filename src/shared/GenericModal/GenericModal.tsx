import { Box, ClickAwayListener } from "@mui/material";
import { SCModal } from "./GenericModal.styles";

type GenericModalProps = {
  children: React.ReactElement;
  open: boolean;
  handleClose: () => void;
};

function GenericModal({ open, handleClose, children }: GenericModalProps) {
  return (
    <SCModal open={open} onClose={handleClose}>
      <Box>
        <ClickAwayListener onClickAway={handleClose}>
          <Box className="modal-content">{children}</Box>
        </ClickAwayListener>
      </Box>
    </SCModal>
  );
}

export default GenericModal;
