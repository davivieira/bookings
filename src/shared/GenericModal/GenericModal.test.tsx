/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import GenericModal from './GenericModal';

jest.mock('./GenericModal.styles', () => ({
  SCModal: jest.fn(({ open, children }: any) => (open ? <div>{children}</div> : null)),
}));

jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  ClickAwayListener: jest.fn(({ children, onClickAway }: any) => (
    <div onClick={() => onClickAway && onClickAway()}>{children}</div>
  )),
  Box: jest.fn(({ children, ...props }: any) => <div {...props}>{children}</div>),
}));

const mockHandleClose = jest.fn();

describe('GenericModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot when open is true', () => {
    const { asFragment } = render(
      <GenericModal open={true} handleClose={mockHandleClose}>
        <div>Modal Content</div>
      </GenericModal>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the modal when open is true', () => {
    render(
      <GenericModal open={true} handleClose={mockHandleClose}>
        <div>Modal Content</div>
      </GenericModal>
    );

    expect(screen.getByText('Modal Content')).toBeInTheDocument();
  });

  it('does not render the modal when open is false', () => {
    render(
      <GenericModal open={false} handleClose={mockHandleClose}>
        <div>Modal Content</div>
      </GenericModal>
    );

    expect(screen.queryByText('Modal Content')).not.toBeInTheDocument();
  });

  it('calls handleClose when clicking away', async () => {
    const user = userEvent.setup();
    render(
      <GenericModal open={true} handleClose={mockHandleClose}>
        <div>Modal Content</div>
      </GenericModal>
    );

    await user.click(screen.getByText('Modal Content').parentElement!);
    expect(mockHandleClose).toHaveBeenCalledTimes(1);
  });

  it('renders children inside the modal', () => {
    render(
      <GenericModal open={true} handleClose={mockHandleClose}>
        <div>Modal Content</div>
      </GenericModal>
    );

    expect(screen.getByText('Modal Content')).toBeInTheDocument();
  });
});
