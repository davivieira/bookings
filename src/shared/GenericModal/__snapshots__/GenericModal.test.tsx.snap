// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericModal matches snapshot when open is true 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <div
          class="modal-content"
        >
          <div>
            Modal Content
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
